src/api/server.rkt

```diff
@@ -53,75 +82,394 @@
+  (place-channel-put receptionist (list 'wait job-id b))
+  (define finished-result (place-channel-get a))
+  (when verbose
+    (eprintf "Done waiting for: ~a\n" job-id))
+  finished-result)
+
+(define (make-receptionist)
+  (place/context*
+   ch
+   #:parameters (*flags* *num-iterations*
+                         *num-points*
+                         *timeout*
+                         *reeval-pts*
+                         *node-limit*
+                         *max-find-range-depth*
+                         *pareto-mode*
+                         *platform-name*
+                         *loose-plugins*)
+   (parameterize ([current-error-port (open-output-nowhere)]) ; hide output
+     (load-herbie-plugins))
+   ; not sure if the above code is actaully needed.
+   (define completed-work (make-hash))
+   (define workers (make-hash))
+   (define waiting (make-hash))
+   (when verbose
+     (eprintf "Receptionist waiting for work.\n"))
+   (for ([i (in-naturals)])
+     ;  (eprintf "Receptionist msg ~a handled\n" i)
+     (match (place-channel-get ch)
+       ; Returns the current count of workers.
+       [(list 'count handler) (place-channel-put handler (hash-count workers))]
+       ; Retreive the improve results for results.json
+       [(list 'improve handler)
+        (define improved-list
+          (for/list ([(job-id result) completed-work]
+                     #:when (equal? (hash-ref result 'command) "improve"))
+            (get-table-data-from-hash result (make-path job-id))))
+        (place-channel-put handler improved-list)]
+       ; Start a worker on a job. Unless the job-id is marked as finished then move to next state.
+       [(list 'start self command job-id)
+        (if (hash-has-key? completed-work job-id)
+            (place-channel-put self (list 'finished job-id (hash-ref completed-work job-id)))
+            (let ([worker (make-worker)])
+              ; Maybe this should be worker-id and we should pre allocate workers based on threads available.
+              (hash-set! workers job-id worker)
+              (when verbose
+                (eprintf "Starting worker [~a] on [~a].\n"
+                         job-id
+                         (test-name (herbie-command-test command))))
+              (place-channel-put worker (list 'apply self command job-id))))]
+       ; Job is finished notified the waiting list with the result of the job.
+       [(list 'finished job-id result)
+        (when verbose
+          (eprintf "Job ~a finished, saving result.\n" job-id))
+        ; Notifed job has been completed, save the result.
+        ; let GC collect worker 🤞.
+        (hash-set! completed-work job-id result)
+        (hash-remove! workers job-id)
+        (define maybe-wait-list (hash-ref waiting job-id #f))
+        (when maybe-wait-list
+          (when verbose
+            (eprintf "waiting job ~a completed\n" job-id))
+          (for ([waiting maybe-wait-list])
+            (place-channel-put waiting result))
```

Comment:
It's fine but I think if you make the default value in `hash-ref` be `'()` then you can just loop over it without the `when` test.

Reply:
Thats clever. Thank you!